storage_controller/src/service.rs

```diff
@@ -3746,15 +3750,58 @@ impl Service {
         )
         .await;
         failpoint_support::sleep_millis_async!("tenant-create-timeline-shared-lock");
-        let create_mode = create_req.mode.clone();
+        let is_import = create_req.is_import();
 
         let timeline_info = self
             .tenant_timeline_create_pageservers(tenant_id, create_req)
             .await?;
 
-        let safekeepers = if safekeepers {
+        let selected_safekeepers = if is_import {
+            let shards = {
+                let locked = self.inner.read().unwrap();
+                locked
+                    .tenants
+                    .range(TenantShardId::tenant_range(tenant_id))
+                    .map(|(ts_id, _)| ts_id.to_index())
+                    .collect::<Vec<_>>()
+            };
+
+            if !shards
+                .iter()
+                .map(|shard_index| shard_index.shard_count)
+                .all_equal()
+            {
+                return Err(ApiError::InternalServerError(anyhow::anyhow!(
+                    "Inconsistent shard count"
+                )));
+            }
+
+            let import = TimelineImport {
+                tenant_id,
+                timeline_id,
+                shard_statuses: ShardImportStatuses::new(shards),
+            };
+
+            let inserted = self
+                .persistence
+                .insert_timeline_import(import.to_persistent())
+                .await
+                .context("timeline import insert")
+                .map_err(ApiError::InternalServerError)?;
```

Comment:
This is running in a separate transaction from the `tenant_timeline_create_pageservers` call above.
This means we will lose track of the import if we crash inbetween above and here.
Can't have that.

If I'm wrong about that, please put an explainer comment on why that is.

Reply:
When thinking about this, please also consider behavior under split brain & storcon deploy.

Reply:
`tenant_timeline_create_pageservers` calls into the pageservers to create the timeline - doesn't touch the persistence layer. If we crash in between, the caller should retry the request, which would eventually insert the import.

Let's say the import is done before we persist it in the db. In that case the pageserver will retry the request forever.

I think it's sound, but rather delicate. Like most of this code path it relies on client retries.

Reply:
Hm ok I missed that for PSes we don't keep track of timelines.

I guess the issue will present itself again once we tend to the creation of the sk timelines for imported timelines, which this PR is still punting on.

(IIRC the implict creation of timelines is slated to go away as part of the sk migrations / membership change project)